src/coreclr/vm/assembly.cpp

```diff
@@ -535,8 +518,7 @@ Assembly *Assembly::CreateDynamic(AssemblyBinder* pBinder, NativeAssemblyNamePar
 
             //Cannot fail after this point
```

Comment:
If something fails between the DomainAssembly constructor and this point, are we going to leak the memory allocated on the `pamTracker` now that we are releasing it much sooner?

Reply:
I was worried about the situation where we do not have per-assembly loader allocator. Ideally, everything that was allocated is released if anything in this method fails.

After closer inspection, I do not think this change is introducing any new issues.

There are pre-existing issues though. For example, if `InitVirtualCallStubManager` above fails, the GC handle allocated at https://github.com/dotnet/runtime/blob/ffd3b187038a107e29cd425de3a748540a75e33e/src/coreclr/vm/loaderallocator.cpp#L1025 is going leak.

Reply:
> I was worried about the situation where we do not have per-assembly loader allocator. Ideally, everything that was allocated is released if anything in this method fails.

It is better to have `AllocMemTracker amTracker` in the top-level scope of the operation for this reason. It makes it easy to see that everything allocated via `AllocMemTracker` is going to be released in case the operation fails. Moving the `AllocMemTracker` into the inner scope does not look like an improvement.

Reply:
> Do you mean making the DomainAssembly constructor take in an AllocMemTracker (instead of it handling creating one specifically for creating the Assembly)? 

Yes.

> I was thinking that it was clearer for the caller to not have to be concerned about creating an AllocMemTracker at all 

It is less correct. If there are any failure point between the `AllocMemTracker.SuppressRelease` call and returning from the QCall, we will have a memory leak on failure (for the non-collectible dynamic assembly). There does not seems to be any such failure points currently, but it is hard to see that. 

Non-collectible DomainAssembly / Assembly cannot free the memory allocated on the loader heaps themselves when there is a failure.